software/base/src/main/java/org/apache/brooklyn/entity/machine/SetHostnameCustomizer.java

```diff
@@ -0,0 +1,201 @@
+            "hostname.local.address",
+            "Host address, as known on the local box. Config is set on the location.");
+    
+    private final ConfigBag config;
+
+    public SetHostnameCustomizer(ConfigBag config) {
+        // TODO Any checks that they've given us sufficient configuration?
+        this.config = config;
+    }
+    
+    @Override
+    public void customize(MachineLocation machine) {
+        String localHostname = setLocalHostname((SshMachineLocation) machine);
+        machine.config().set(LOCAL_HOSTNAME, localHostname);
+
+        String localIp = execHostnameMinusI((SshMachineLocation) machine);
+        machine.config().set(LOCAL_IP, localIp);
+    }
+
+    protected String generateHostname(SshMachineLocation machine) {
+        String hostnameTemplate = config.get(HOSTNAME_TEMPLATE);
+        if (Strings.isNonBlank(hostnameTemplate)) {
+            return TemplateProcessor.processTemplateContents(hostnameTemplate, machine, ImmutableMap.<String, Object>of());
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Sets the machine's hostname to the value controlled by fixed_hostname and hostname_template. 
+     * If these are blank (and fixed_domain is blank), then just return the current hostname of 
+     * the machine.
+     */
+    public String setLocalHostname(SshMachineLocation machine) {
+        String hostFixed = config.get(FIXED_HOSTNAME);
+        String domainFixed = config.get(FIXED_DOMAIN);
+        String hostnameTemplate = config.get(HOSTNAME_TEMPLATE);
+        
+        String hostname;
+        if (Strings.isNonBlank(hostFixed)) {
+            hostname = hostFixed;
+        } else {
+            if (Strings.isNonBlank(hostnameTemplate)) {
+                hostname = generateHostname(machine);
+            } else {
+                hostname = execHostname(machine);
+                if (Strings.isBlank(domainFixed)) {
+                    return hostname;
+                }
+            }
+        }
+        
+        return setLocalHostname(machine, hostname, domainFixed);
+    }
+    
+    /**
+     * Sets the machine's hostname to the given value, ensuring /etc/hosts and /etc/sysconfig/network are both 
+     * correctly updated.
+     */
+    public String setLocalHostname(SshMachineLocation machine, String hostName, String domainFixed) {
+        log.info("Setting local hostname of " + machine + " to " + hostName 
+                + (Strings.isNonBlank(domainFixed) ? ", " + domainFixed : ""));
+        
+        boolean hasDomain = Strings.isNonBlank(domainFixed);
+        String fqdn = hasDomain ? hostName+"."+domainFixed : hostName;
+        
+        exec(machine, true, 
+                BashCommands.sudo(String.format("sed -i.bak -e '1i127.0.0.1 %s %s' -e '/^127.0.0.1/d' /etc/hosts", fqdn, hostName)),
+                BashCommands.sudo(String.format("sed -i.bak -e 's/^HOSTNAME=.*$/HOSTNAME=%s/' /etc/sysconfig/network", fqdn)),
```

Comment:
Agreed. This is focused on CentOS and RHEL just now (there's a comment in the javadoc). We should at a later stage add support for other distros.